lib/queue_classic.rb

```diff
@@ -50,7 +50,8 @@ def self.has_connection?
 
   def self.default_conn_adapter
     t = Thread.current
-    return t[:qc_conn_adapter] if t[:qc_conn_adapter]
+    return t[:qc_conn_adapter] if t[:qc_conn_adapter] && t[:qc_conn_adapter].active?
```

Comment:
I think connection should be closed here if it's not active due to `PQTRANS_INERROR`. I.e. `t[:qc_conn_adapter].disconnect`. It handles disconnection errors inside but logs them, and that can be a problem if trying to disconnect already disconnected connection raises exception (I'm not sure if it does). So maybe separate disconnection method will be needed, or `t[:qc_conn_adapter].connection.close rescue nil`.

For example, [activerecord uses `rescue nil`](https://github.com/rails/rails/blob/cadc5de223c4de30cc8280dbb335d02b07c0c44f/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L295-L302) so [when fetching connection from pool it can be safely closed on error](https://github.com/rails/rails/blob/cadc5de223c4de30cc8280dbb335d02b07c0c44f/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L859-L868).

Otherwise, connection will stay open until GC finalizes its object (which can take long time, GC is not designed for connection lifetime management).